### PR TITLE
[FLINK-20961][table-planner-blink] Fix NPE when no assigned timestamp defined

### PR DESCRIPTION
**Pull request by YuvalItzchakov**
## What is the purpose of the change

Fix an erroneous NPE created when using a table from a DataStream that did not assign timestamps and watermarks

## Brief change log

- Fix direct access to `ctx.timestamp()`, validate that it has a timestamp before

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

(Maybe, not sure)  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? No

### CODE REVIEW

**Review by wuchong — COMMENTED** [review]
Please also add a test to cover this.

**Comment by YuvalItzchakov** [comment]
@wuchong Where would be the proper place to test this?

**Comment by wuchong** [comment]
Hi @YuvalItzchakov , I think you can add an IT case in `TimeAttributesITCase`, you can create a `DataStream` without `assignTimestampsAndWatermarks` which should reproduce this problem.

Btw, please also fix the checkstyle error:

```
2021-01-13T13:49:30.9144714Z [INFO] --- scalastyle-maven-plugin:1.0.0:check (default) @ flink-table-planner-blink_2.11 ---
2021-01-13T13:49:43.7187879Z error file=/__w/3/s/flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala message=File line length exceeds 100 characters line=521
```

**Comment by wuchong** [comment]
Btw, `TimeAttributesITCase` is only located in old planner, we can create one under `org.apache.flink.table.planner.runtime.stream.table` package in blink planner, and copy tests to it.

**Comment by wuchong** [comment]
Hi @YuvalItzchakov  do you have to to update the pull request?

**Comment by YuvalItzchakov** [comment]
Hi @wuchong. Yes I'm working on it ATM, struggling a bit to get Flink to compile from source but I'm confident I'll resolve it soon

**Comment by YuvalItzchakov** [comment]
@flinkbot run azure

**Comment by YuvalItzchakov** [comment]
@wuchong Can I re-run the CI tests?

**Review by wuchong — APPROVED** [review]
LGTM.

**Review by wuchong — CHANGES_REQUESTED** [review]
Could you check the failed tests? 

`MatchRecognizeITCase.testWindowedGroupingAppliedToMatchRecognize`

**Comment by YuvalItzchakov** [comment]
@wuchong Apparently not all interfaces of `Context` support `hasTimestamp`. I refactored the code to just do a `null` check on the returned timestamp.

**Comment by YuvalItzchakov** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala [hunk 1]
```diff
@@ -512,18 +512,21 @@ object GenerateUtils {
...
 
     val accessCode =
       s"""
-         |$resultTerm = $TIMESTAMP_DATA.fromEpochMillis($contextTerm.timestamp());
-         |if ($resultTerm == null) {
-         |  throw new RuntimeException("Rowtime timestamp is null. Please make sure that a " +
-         |    "proper TimestampAssigner is defined and the stream environment uses the EventTime " +
-         |    "time characteristic.");
+         |$timestamp = $contextTerm.timestamp();
+         |if ($timestamp == null) {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:524]: I simple way for this problem can be `if ($contextTerm.timestamp() == null)`, then you don't need to declare a tmp variable.

Comment (YuvalItzchakov) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:524]: @wuchong Right, it's either this or retrieving the timestamp twice. Whichever is preferred.

Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:524]: The cost of get timestamp is negligible. Therefore, I prefer the simper way.

Comment (YuvalItzchakov) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:524]: Done

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala [hunk 2]
```diff
@@ -909,6 +908,62 @@ object GenerateUtils {
...
+    })
+    compares.mkString
+  }
+
+  /**
+   * Generates code for comparing row keys.
+   * TODO: Remove this after [[MultiFieldRangeBoundComparatorCodeGenerator]]
+   * is refactored using SortSpec.
+   */
+  def generateRowCompare(
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:920]: It seems this method is not used. Is it introduced by accident?

Comment (YuvalItzchakov) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:920]: Weird, I did not add that. Will revert.

Comment (YuvalItzchakov) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:920]: @wuchong Did you rebase off master? Diffing master I see this was added.

**Comment by YuvalItzchakov** [comment]
@wuchong I'm going to re-open this PR under a fresh branch.

**Pull request closed without merging**